### PR TITLE
feat(slang): complete enums and user-defined value types

### PR DESCRIPTION
**Pull request by hedgar2017**
Completes enums and user-defined value types on the Slang frontend:

- An enum member `E.Variant`, its qualified form `C.E.Variant`, and `type(E).min` / `type(E).max` materialize as the member's ordinal: a `sol.constant` at the EVM word, cast by `sol.enum_cast` to `!sol.enum<N>`; the ordinal is a member's position among the enum's declarations, and the maximum is read back off the dialect type that already bounds it.
- `E(v)` and `uint8(e)` lower to `sol.enum_cast` in either direction, a wider target chaining `sol.cast` behind it; the range check on an integer narrowed into an enum belongs to the lowering pass, not the frontend.
- An enum comparison emits `sol.cmp` at `!sol.enum<N>` with no widening in front of it, the dialect admitting an enum wherever it admits an integer.
- An enum-typed state variable, mapping key, struct field, or array element carries `!sol.enum<N>` through `sol.state_var`, `sol.mapping`, `sol.gep`, and `sol.load`; an uninitialized local, a named return, and `delete` all default-initialize to ordinal zero through the same constant-and-cast pair solc emits.
- `T.wrap(x)` and `T.unwrap(v)` emit nothing, a user-defined value type lowering to the type it wraps, so both are the identity case of an explicit conversion.
- A user-defined operator — the fourteen binary forms, `~`, and unary `-` — lowers to `sol.call` of the function a `using {f as op} for T global;` directive binds it to, which the binder resolves and the AST reads back; being a call, its operands evaluate left-first rather than in a built-in operator's right-first order.
- The operator-bound free functions emit into the contract body alongside the contract's own, MLIR having no file scope for them to land in.

### CODE REVIEW

File: solx-slang/src/source_unit.rs [hunk 1]
```diff
@@ -44,4 +50,29 @@ impl<'context> SourceUnitScope<'context> {
...
+                        UsingClause::IdentifierPath(_) => None,
+                    }
+                }
+                _ => None,
+            })
+            .flat_map(|deconstruction| deconstruction.symbols().iter().collect::<Vec<_>>())
+            .filter(|symbol| symbol.alias().is_some())
+            .filter_map(|symbol| match symbol.name().resolve_to_definition()? {
+                Definition::Function(function) => Some(function),
+                _ => None,
```
Comment (PavelKopyl) [inline solx-slang/src/source_unit.rs:74]: I wonder is there a chance to get a symbol collision between an operator function and a same signature contract member, as we do not mangle operator functions: 

```
using {add as +} for T global;

// File-level free function, bound as the `+` operator for T.
function add(T a, T b) pure returns (T) {
    return T.wrap(T.unwrap(a) + T.unwrap(b));
}

contract C {
    // Legal Solidity: a member function with the same name and parameter
    // types as the free function. Inside C it shadows the free `add` for
    // direct calls, but `+` still binds the free one.
    function add(T a, T b) public pure returns (T) {
        return a + b; // lowers to sol.call of the *free* add
    }
}
```
?

File: solx-mlir/tests/lit/type_casts.sol [hunk 2]
```diff
@@ -86,4 +98,16 @@ contract C {
     function array_to_memory(uint256[] calldata source) external pure returns (uint256[] memory) {
         return uint256[](source);
     }
+
+    function uint8_to_enum(uint8 x) public pure returns (E) {
+        return E(x);
+    }
```
Comment (PavelKopyl) [inline solx-mlir/tests/lit/type_casts.sol:104]: It's reasonable to add a runtime test (if such does't exist yet) to check when 'x' is out of range for the enum 'E'. This should cause a runtime Panic.  On the other hand we have such a test in era-solidity.

File: solx-slang/src/source_unit.rs [hunk 3]
```diff
@@ -44,4 +50,29 @@ impl<'context> SourceUnitScope<'context> {
             Contract::new_mlir(mlir, method_identifiers),
         )]))
     }
+
+    /// The functions `unit`'s `using {f as op} for T global;` directives bind operators to, in
+    /// source order and without repeats. A binding reaches the whole compilation unit, so a
+    /// function bound from another file is missing here.
```
Comment (PavelKopyl) [inline solx-slang/src/source_unit.rs:56]: Let's add a TODO here: we don't yet support cross-file binding of operator functions. Since `global` is specifically what makes a `using ... for` attachment visible to importing files, that's behaviour the keyword's semantics require, and we'll need to handle it.

File: solx-mlir/tests/lit/type_casts.sol (hunk 2 above)
Comment (hedgar2017) [reply solx-mlir/tests/lit/type_casts.sol:104]: If solx-solidity already has one, I don't see any reason to add it in this repo.

File: solx-slang/src/source_unit.rs (hunk 1 above)
Comment (hedgar2017) [reply solx-slang/src/source_unit.rs:74]: Yes, tracked in https://github.com/NomicFoundation/solx/issues/560 and fixed in #621 🚀

File: solx-slang/src/source_unit.rs (hunk 3 above)
Comment (hedgar2017) [reply solx-slang/src/source_unit.rs:56]: Good idea, adding!

**Review by PavelKopyl — APPROVED** [review]
LGTM, thanks!
